staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval_test.go

```diff
@@ -0,0 +1,496 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cacher
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"testing"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+)
+
+func intervalFromEvents(events []*watchCacheEvent) *watchCacheInterval {
+	startIndex, endIndex, locker := 0, len(events), &sync.Mutex{}
+	indexer := func(i int) *watchCacheEvent {
+		if len(events) == 0 {
+			return nil
+		}
+		return events[i]
+	}
+	indexValidator := func(_ int) bool { return true }
+	wci := newCacheInterval(startIndex, endIndex, indexer, indexValidator, locker)
+
+	return wci
+}
+
+func bufferFromEvents(events []*watchCacheEvent) *watchCacheIntervalBuffer {
+	wcib := &watchCacheIntervalBuffer{
+		buffer:     make([]*watchCacheEvent, bufferSize),
+		startIndex: 0,
+		endIndex:   len(events),
+	}
+	copy(wcib.buffer, events)
+
+	return wcib
+}
+
+func generateEvents(start, end int) []*watchCacheEvent {
+	n := end - start
+	events := make([]*watchCacheEvent, n)
+	for i := 0; i < n; i++ {
+		events[i] = &watchCacheEvent{
+			Type: watch.Added,
+			Object: makeTestPod(
+				fmt.Sprintf("pod%d", start+i+1),
+				uint64(start+i+1),
+			),
+		}
+	}
+	return events
+}
+
+func verifyEvent(
+	t *testing.T,
+	event, expectedEvent *watchCacheEvent,
+	received, expected int,
+	err error,
+) bool {
```

Comment:
Let's change the semantic of this function to:
- not take testing.T as an argument
- instead of returning bool, return an error (if nil, then it means everything was fine).